server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNode.java

```diff
@@ -226,4 +251,88 @@ public String toString() {
             + ']';
     }
 
+    public record Processors(float min, float max) implements Writeable, ToXContentObject {
+
+        private static final Version MIN_MAX_PROCESSORS_VERSION = Version.V_8_3_0;
+
+        private static final ParseField MIN_FIELD = new ParseField("min");
+        private static final ParseField MAX_FIELD = new ParseField("max");
+
+        public static final ConstructingObjectParser<Processors, String> PROCESSORS_PARSER = new ConstructingObjectParser<>(
+            "processors",
+            false,
+            (args, name) -> new Processors((float) args[0], (float) args[1])
+        );
+
+        static {
+            PROCESSORS_PARSER.declareFloat(ConstructingObjectParser.constructorArg(), MIN_FIELD);
+            PROCESSORS_PARSER.declareFloat(ConstructingObjectParser.constructorArg(), MAX_FIELD);
+        }
+
+        public Processors(int processors) {
+            this((float) processors, (float) processors);
+        }
+
+        static Processors fromXContent(XContentParser parser) throws IOException {
+            if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+                return PROCESSORS_PARSER.parse(parser, null);
+            } else {
+                // For BWC with nodes pre 8.3
+                float processors = parser.floatValue();
+                return new Processors(processors, processors);
+            }
+        }
+
+        public Processors {
+            if (min > max) {
+                throw new IllegalArgumentException(
+                    "min processors must be less than or equal to max processors and it was: min: " + min + " max: " + max
+                );
+            }
+
+            if (min <= 0) {
+                throw new IllegalArgumentException("min processors must be greater than 0, but got " + min);
+            }
+
+            if (max <= 0) {
+                throw new IllegalArgumentException("min processors must be greater than 0, but got " + max);
+            }
+        }
+
+        private static Processors readFrom(StreamInput in) throws IOException {
+            if (in.getVersion().onOrAfter(MIN_MAX_PROCESSORS_VERSION)) {
+                return new Processors(in.readFloat(), in.readFloat());
+            } else {
+                int processors = in.readInt();
+                return new Processors((float) processors, (float) processors);
+            }
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            if (out.getVersion().onOrAfter(MIN_MAX_PROCESSORS_VERSION)) {
+                out.writeFloat(min);
+                out.writeFloat(max);
+            } else {
+                out.writeInt(roundedMin());
+            }
+        }
+
+        public int roundedMin() {
+            return Math.round(min);
```

Comment:
I wonder if we should round this down and round max up? That seems more intuitive, since then the actual quota of processors is still within the range.